geps/gep-2643/index.md

```diff
@@ -0,0 +1,263 @@
+* A backend that is TLS based but not HTTP based (e.g., a Kafka service, or a
+Postgres service, with its listener being TLS enabled).
+* Some WebRTC solutions.
+* Backends that can require direct client-certificate authentication (e.g., OAuth).
+
+For the example cases above, it is desired that the routing is made as a passthrough
+mode, where the `Gateway` passes the packets to the backend without terminating TLS.
+
+While this routing is possible using other mechanisms such as a `TCPRoute` or a
+Kubernetes service of type `LoadBalancer`, it may be desired to have a single point
+of traffic convergence (like one single `LoadBalancer`) for reasons like cost,
+traffic control, etc.
+
+An implementation of `TLSRoute` achieves this end using the
+[server_name TLS attribute](https://datatracker.ietf.org/doc/html/rfc6066#section-3)
+to determine what backend should be used for a given request. `TLSRoute` thereby
+enables the use of a single gateway listener to handle traffic for multiple routes.
+
+## API
+
+```golang
+type TLSRouteSpec struct {
+  // Hostnames defines a set of SNI hostnames that should match against the
+  // SNI attribute of TLS ClientHello message in TLS handshake. This matches
+  // the RFC 1123 definition of a hostname with 2 notable exceptions:
+  //
+  // 1. IPs are not allowed in SNI hostnames per RFC 6066.
+  // 2. A hostname may be prefixed with a wildcard label (`*.`). The wildcard
+  //    label must appear by itself as the first label.
+  //
+  // If a hostname is specified by both the Listener and TLSRoute, there
+  // must be at least one intersecting hostname for the TLSRoute to be
+  // attached to the Listener. For example:
+  //
+  // * A Listener with `test.example.com` as the hostname matches TLSRoutes
+  //   that have specified at least one of `test.example.com` or
+  //   `*.example.com`.
+  // * A Listener with `*.example.com` as the hostname matches TLSRoutes
+  //   that have specified at least one hostname that matches the Listener
+  //   hostname. For example, `test.example.com` and `*.example.com` would both
+  //   match. On the other hand, `example.com` and `test.example.net` would not
+  //   match.
+  //
+  // If both the Listener and TLSRoute have specified hostnames, any
+  // TLSRoute hostnames that do not match the Listener hostname MUST be
+  // ignored. For example, if a Listener specified `*.example.com`, and the
+  // TLSRoute specified `test.example.com` and `test.example.net`,
+  // `test.example.net` must not be considered for a match.
+  //
+  // If both the Listener and TLSRoute have specified hostnames, and none
+  // match with the criteria above, then the TLSRoute is not accepted. The
+  // implementation must raise an 'Accepted' Condition with a status of
+  // `False` in the corresponding RouteParentStatus.
+  // +required
+  // +kubebuilder:validation:MinItems=1
+  // +kubebuilder:validation:MaxItems=16
+  Hostnames []Hostname `json:"hostnames,omitempty"`
```

Comment:
This is what exists on the current API, but IMO the hostnames should be part of a `TLSRouteRule` and not its own field on the spec.

It doesn't make much sense that the rules are an array, that contain an array of backendRefs, but the hostnames are outside of it, but maybe there's some more context here.

Reply:
Synced to @youngnick and @howardjohn and although matching SNI under TLSRouteRule has advantages, moving `hostnames` would be a substantial breaking change for many customers using TLSRoutes today, and using conversion webhooks isn't really under consideration. 

So we're leaning toward promoting `TLSRoute.hostnames[]` to standard now, and monitoring adoption to collect use cases and determine if there is a need for SNI-based fan-out.